### PR TITLE
Feature/2017 devoxx pl Sam Newman

### PR DESCRIPTION
**Pull request by JMesens**

### CODE REVIEW

File: _posts/2017-06-21-Devoxx-pl.md [hunk 1]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
 
 Extended blog coming later.
 
+## Feature Branches And Toggles In A Post-GitHub World ([Sam Newman](https://twitter.com/@samnewman))
+
+Sam told us about his experience at a project that was having trouble merging branches.
+The release branch for next release was called R3, but for a large refactoring branch R4 was created.
```
Comment (bart-blommaerts) [inline _posts/2017-06-21-Devoxx-pl.md]: for **the** next release?

Comment (bart-blommaerts) [inline _posts/2017-06-21-Devoxx-pl.md]: but for a large refactoring**,** branch

File: _posts/2017-06-21-Devoxx-pl.md [hunk 2]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+>A flag should be set and evaluated in as few places as possible, preferably once each.
+>Flags should be removed when the new implementation is done.
+
+More info: [Trunk-based development](https://trunkbaseddevelopment.com/)
+
+Changes to existing functionality can be done by providing an abstraction above the existing functionality.
+The new functionality can then be developed for the abstraction and when it is done, change to the new implementation.
+**Branch-by abstraction** has the side-benefit that it can be used for **A/B and canary releasing**.
+
+The Continuous Delivery book tells us to treat every check in as a possible release candidate.
```
Comment (bart-blommaerts) [inline _posts/2017-06-21-Devoxx-pl.md]: https://www.amazon.de/Continuous-Delivery-Deployment-Automation-Addison-Wesley/dp/0321601912

File: _posts/2017-06-21-Devoxx-pl.md [hunk 3]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+Developers start with the assumption it is worthy, CI tool decides if it is.
+Deploy frequently with small changes, makes easier to rollback, lowering the chance of problems.
+
+### Git
+
+And then there was **Git**, developed by Linus Torvalds with the goal to merge a patch in less than three seconds.
```
Comment (bart-blommaerts) [inline _posts/2017-06-21-Devoxx-pl.md]: https://en.wikipedia.org/wiki/Linus_Torvalds

File: _posts/2017-06-21-Devoxx-pl.md [hunk 4]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+Branches where much more lightweight and every repository contained the full source history.
```
Comment (bart-blommaerts) [inline _posts/2017-06-21-Devoxx-pl.md]: In Git, branches are much more ... and every repository contains ...

File: _posts/2017-06-21-Devoxx-pl.md [hunk 5]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+
+In 2008 **[GitHub](https://github.com)** was founded and introduced pull requests.
+If you wanted to contribute to open source projects before pull requests you had to:
+* develop it locally, 
+* generate patch file
```
Comment (bart-blommaerts) [inline _posts/2017-06-21-Devoxx-pl.md]: generate a patch file

File: _posts/2017-06-21-Devoxx-pl.md [hunk 6]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+Branches where much more lightweight and every repository contained the full source history.
+
+In 2008 **[GitHub](https://github.com)** was founded and introduced pull requests.
+If you wanted to contribute to open source projects before pull requests you had to:
+* develop it locally, 
+* generate patch file 
+* mail it over to the project owners.
+This feature contributed to GitHub's success, three years later in 2011 they passed SourceForge and Google Code in popularity.
+
+Sam made the remark that pull requests used branches, which might bring problems.
```
Comment (bart-blommaerts) [inline _posts/2017-06-21-Devoxx-pl.md]: request use branches

they still do :)

File: _posts/2017-06-21-Devoxx-pl.md [hunk 7]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+If you wanted to contribute to open source projects before pull requests you had to:
+* develop it locally, 
+* generate patch file 
+* mail it over to the project owners.
+This feature contributed to GitHub's success, three years later in 2011 they passed SourceForge and Google Code in popularity.
+
+Sam made the remark that pull requests used branches, which might bring problems. 
+On top of that **[GitFlow](https://www.atlassian.com/git/tutorials/comparing-workflows#gitflow-workflow)** was introduced.
+Because GitFlow introduces even more branches it is in controversy with fast deployment and small changes cycle.
+With tools like [split.io](https://www.split.io) and [launch darkly](https://launchdarkly.com) Git-Flow is not needed, if merged frequently.
```
Comment (bart-blommaerts) [inline _posts/2017-06-21-Devoxx-pl.md]: GitFlow

File: _posts/2017-06-21-Devoxx-pl.md [hunk 8]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+
+To concluded Sam asked himself if branches where evil.
```
Comment (bart-blommaerts) [inline _posts/2017-06-21-Devoxx-pl.md]: To conclude, Sam challenged to audience and asked if branches are evil.

File: _posts/2017-06-21-Devoxx-pl.md [hunk 9]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+No, experimental and release branches that are not going te merged have its uses.
```
Comment (bart-blommaerts) [inline _posts/2017-06-21-Devoxx-pl.md]: that might never get merged, have their uses.

File: _posts/2017-06-21-Devoxx-pl.md (hunk 8 above)
Comment (JMesens) [inline _posts/2017-06-21-Devoxx-pl.md]: You probably meant: To conclude, Sam challenged the audience and asked if branches are evil.

File: _posts/2017-06-21-Devoxx-pl.md [hunk 10]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
 
 Extended blog coming later.
 
+## Feature Branches And Toggles In A Post-GitHub World ([Sam Newman](https://twitter.com/@samnewman))
+
+Sam told us about his experience at a project that was having trouble merging branches.
+The release branch for the next release was called R3, but for a large refactoring, branch R4 was created.
+Afterwards he described merging the branches as a car crash.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: Afterwards, ...
(comma)

File: _posts/2017-06-21-Devoxx-pl.md [hunk 11]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+They even needed to introduce a dedicated R3-R4 merge bug fix team.
+Then **Continuous Integration** was meant to prevent the merging problem.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: Would rephrase:
**Continuous Integration** had been set up in order to prevent the merging issues.

File: _posts/2017-06-21-Devoxx-pl.md [hunk 12]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+Developers had to check in their changes and the CI tool would validate the code.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: Would rephrase:
The code, pushed by the developers, would get automatically validated by the CI setup.

File: _posts/2017-06-21-Devoxx-pl.md [hunk 13]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+The problem with R3-R4 release was that validation was done only for a branch, not on the integrated branches.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: ... with **the** R3-R4 release ...
... branch **and** not on the integrated branches.

File: _posts/2017-06-21-Devoxx-pl.md [hunk 14]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+
+>The integration should be validated every day and when the build breaks, fix it!
+
+For unfinished work we can wait till it is ready before checking in.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: For unfinished work, we can wait **until** ...
(comma)

File: _posts/2017-06-21-Devoxx-pl.md [hunk 15]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+This exposes us to the risk of losing work when it's only on the developers computer.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: developer's

File: _posts/2017-06-21-Devoxx-pl.md [hunk 16]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+Merging branches can be a difficult task and might lead to commit race, to offload the effort to a colleague.
+
+### Trunk-based development
+
+A third option would be to 'check in anyway', called **trunk-based development**.
+Every commit integrates to the trunk and developers should integrate their local changes daily.
+Small changes and integrating often makes it easier to merge new code.
+
+New half-finished features can be hidden with **feature toggles**.
+These toggles can me managed using flags or configurations (eg, in Zookeeper, Consul, ...).
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: ... can **be** managed ...

File: _posts/2017-06-21-Devoxx-pl.md [hunk 17]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+>A flag should be set and evaluated in as few places as possible, preferably once each.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: Probably sounds better:
... preferably **only** once.

File: _posts/2017-06-21-Devoxx-pl.md [hunk 18]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+>Flags should be removed when the new implementation is done.
+
+More info: [Trunk-based development](https://trunkbaseddevelopment.com/)
+
+Changes to existing functionality can be done by providing an abstraction above the existing functionality.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: Changes to **an** existing ...

File: _posts/2017-06-21-Devoxx-pl.md [hunk 19]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+The new functionality can then be developed for the abstraction and when it is done, change to the new implementation.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: ... **changed** to the new implementation.

File: _posts/2017-06-21-Devoxx-pl.md [hunk 20]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+**Branch-by abstraction** has the side-benefit that it can be used for **A/B and canary releasing**.
+
+[The Continuous Delivery](https://www.amazon.de/Continuous-Delivery-Deployment-Automation-Addison-Wesley/dp/0321601912) book tells us to treat every check in as a possible release candidate.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: Think that in this case it's written as **check-in**.

File: _posts/2017-06-21-Devoxx-pl.md [hunk 21]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+Developers start with the assumption it is worthy, CI tool decides if it is.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: ... assumption **that** it is worthy **after which the CI tool decides whether it truly is**.

File: _posts/2017-06-21-Devoxx-pl.md [hunk 22]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+Deploy frequently with small changes, makes easier to rollback, lowering the chance of problems.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: ... makes **it** easier to rollback **and lowers the risk of running into problems**.

File: _posts/2017-06-21-Devoxx-pl.md [hunk 23]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+
+### Git
+
+And then there was **Git**, developed by [Linus Torvalds](https://en.wikipedia.org/wiki/Linus_Torvalds) with the goal to merge a patch in less than three seconds.
+In Git branches where much more lightweight and every repository contains the full source history.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: I assume you mean local repository?
In Git, branches **are** much more lightweight and every **local** repository contains the full source history.

File: _posts/2017-06-21-Devoxx-pl.md [hunk 24]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+
+In 2008 **[GitHub](https://github.com)** was founded and introduced pull requests.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: In 2008,
(comma)

File: _posts/2017-06-21-Devoxx-pl.md [hunk 25]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+If you wanted to contribute to open source projects before pull requests you had to:
+* develop it locally,
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: May need to prepend this with a newline, otherwise the list is not properly rendered.
Verify if it's the case, if not, you can ignore this comment.

File: _posts/2017-06-21-Devoxx-pl.md [hunk 26]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+
+And then there was **Git**, developed by [Linus Torvalds](https://en.wikipedia.org/wiki/Linus_Torvalds) with the goal to merge a patch in less than three seconds.
+In Git branches where much more lightweight and every repository contains the full source history.
+
+In 2008 **[GitHub](https://github.com)** was founded and introduced pull requests.
+If you wanted to contribute to open source projects before pull requests you had to:
+* develop it locally, 
+* generate a patch file 
+* mail it over to the project owners.
+This feature contributed to GitHub's success, three years later in 2011 they passed SourceForge and Google Code in popularity.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: May also need to prepend this line with a newline.

File: _posts/2017-06-21-Devoxx-pl.md [hunk 27]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+
+Sam made the remark that pull requests use branches, which might bring problems. 
+On top of that **[GitFlow](https://www.atlassian.com/git/tutorials/comparing-workflows#gitflow-workflow)** was introduced.
+Because GitFlow introduces even more branches it is in controversy with fast deployment and small changes cycle.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: ... even more branches, it is ...
(comma)

File: _posts/2017-06-21-Devoxx-pl.md [hunk 28]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+With tools like [split.io](https://www.split.io) and [launch darkly](https://launchdarkly.com) GitFlow is not needed, if merged frequently.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: Split
LaunchDarkly
, GitFlow is not ... (comma)

File: _posts/2017-06-21-Devoxx-pl.md [hunk 29]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+
+To conclude, Sam challenged the audience and asked if branches are evil.
+No, experimental and release branches that might never get merged, have their uses.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: Would rephrase:
The conclusion was that experimental and release branches, that might even never get merged, still have their uses.

File: _posts/2017-06-21-Devoxx-pl.md [hunk 30]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+Pull request also works good for open source.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: Would rephrase:
The pull request mechanism works well in open source projects.

File: _posts/2017-06-21-Devoxx-pl.md [hunk 31]
```diff
@@ -105,10 +105,74 @@ To build up this discipline, automated verification can be seen as the software
...
+Except for these cases he recommends to prevent branches and to **keep batch sizes small, integrate often and ship often**.
```
Comment (yannickdeturck) [inline _posts/2017-06-21-Devoxx-pl.md]: Which cases?

File: _posts/2017-06-21-Devoxx-pl.md (hunk 25 above)
Comment (JMesens) [inline _posts/2017-06-21-Devoxx-pl.md]: It seems to be fine.

File: _posts/2017-06-21-Devoxx-pl.md (hunk 26 above)
Comment (JMesens) [inline _posts/2017-06-21-Devoxx-pl.md]: Indeed!

File: _posts/2017-06-21-Devoxx-pl.md (hunk 31 above)
Comment (JMesens) [inline _posts/2017-06-21-Devoxx-pl.md]: The ones above: experimental, release branches and pull requests.
